tsl/src/gapfill/exec.c

```diff
@@ -280,53 +520,38 @@ gapfill_begin(CustomScanState *node, EState *estate, int eflags)
 
 	arg_value = gapfill_exec_expr(state, linitial(func->args), &isnull);
 	if (isnull)
-	{
 		ereport(ERROR,
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 				 errmsg("invalid time_bucket_gapfill argument: bucket_width cannot be NULL")));
-	}
+
 	state->gapfill_period = gapfill_period_get_internal(func->funcresulttype,
 														exprType(linitial(func->args)),
 														arg_value);
 
-	/* check if gapfill start was left out */
+	/*
+	 * check if gapfill start was left out so we have to infer from WHERE
+	 * clause
+	 */
 	if (is_const_null(lthird(func->args)))
-		ereport(ERROR,
-				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-				 errmsg("invalid time_bucket_gapfill argument: start cannot be NULL")));
-	else
 	{
-		if (!is_simple_expr(lthird(func->args)))
-			ereport(ERROR,
-					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-					 errmsg("invalid time_bucket_gapfill argument: start must be a simple "
-							"expression")));
+		int64 start = infer_gapfill_boundary(state, GAPFILL_START);
+		Const *expr = make_const_value(state->gapfill_typid, start);
 
+		state->gapfill_start = align_with_time_bucket(state, (Expr *) expr);
+	}
+	else
+	{
 		/*
 		 * pass gapfill start through time_bucket so it is aligned with bucket
 		 * start
 		 */
-		call->args = list_make2(linitial(func->args), lthird(func->args));
-		arg_value = gapfill_exec_expr(state, (Expr *) call, &isnull);
-
-		/*
-		 * the default value for start is NULL but this is checked above, when
-		 * a non-Const is passed here that evaluates to NULL we bail
-		 */
-		if (isnull)
-			ereport(ERROR,
-					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-					 errmsg("invalid time_bucket_gapfill argument: start cannot be NULL")));
-
-		state->gapfill_start = gapfill_datum_get_internal(arg_value, func->funcresulttype);
+		state->gapfill_start = align_with_time_bucket(state, lthird(func->args));
 	}
 	state->next_timestamp = state->gapfill_start;
 
 	/* gap fill end */
 	if (is_const_null(lfourth(func->args)))
-		ereport(ERROR,
-				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-				 errmsg("invalid time_bucket_gapfill argument: finish cannot be NULL")));
+		state->gapfill_end = infer_gapfill_boundary(state, GAPFILL_END);
```

Comment:
why do we not need to align start and not end btw?

Reply:
because `end` is only used to check whether we have reached it while `start` is also used to initialize generated buckets